### PR TITLE
server: support deprecation of EOF packets

### LINKED ISSUE #32118: server: support `CLIENT_DEPRECATE_EOF` capability flag
## Enhancement

MySQL deprecated `EOF` packets since MySQL v5.7.5 and [`CLIENT_DEPRECATE_EOF`](https://dev.mysql.com/doc/internals/en/capability-flags.html) was introduced since then.

This flag, as well as the deprecation of EOF was not supported by TiDB. This is fine since this flag requires a server-client negotiation according to the MySQL protocol. All MySQL clients are able to communicate with TiDB by treating it as 'the old MySQL server'.

However, it is better to make it implemented for:
* EOF has been deprecated for a long time, we hope that TiDB could be 'up-to-date' for MySQL compatibilities.
* The 'old server' might not be fully tested for the new versions of MySQL clients. For example: [Connector/J bug #10625](https://bugs.mysql.com/bug.php?id=106252)

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Adds support for `CLIENT_DEPRECATE_EOF`

Issue Number: close #32118

Note that  #32119 has a different solution to the same problem

Related: https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_response_packets.html

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support for the CLIENT_DEPRECATE_EOF feature of the MySQL protocol was added.
```

We could also say something about improved support for "MySQL Connector/J", but this may impact more than just the connector for Java.

### CODE REVIEW

**Comment by dveeden** [comment]
I tried to trigger the issue with Connector/J that was reported, but wasn't able to reproduce the original issue.
```java
package com.pingcap.tidb_demo_java;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.ResultSet;

public class App 
{
    public static void main( String[] args )
    {
	try {
		Class.forName("com.mysql.cj.jdbc.Driver").newInstance();
	} catch (Exception ex) {
		System.out.println("Failed to load driver: " + ex.getMessage());
	}

	try {
		Connection conn = DriverManager.getConnection("jdbc:mysql:loadbalance://127.0.0.1:4000,127.0.0.1:4001?user=root&useServerPrepStmts=true");
		Statement stmt = conn.createStatement();
		ResultSet rs = stmt.executeQuery("SELECT @@hostname, @@port");
		rs.next();
		System.out.println("Connected to: " + rs.getString("@@hostname") + ":" + rs.getString("@@port"));

		ResultSet rs2 = stmt.executeQuery("SELECT 'foo' a, 'bar' b, REPEAT('x', 100000) c UNION ALL SELECT 'foo', 'bar', REPEAT('y', 10000000)");
		rs2.next();
		System.out.println("Row 1, Column a: " + rs2.getString("a"));
		System.out.println("Row 1, Column b: " + rs2.getString("b"));
		System.out.println("Row 1, Column c length: " + rs2.getString("c").length());

		rs2.next();
		System.out.println("Row 2, Column a: " + rs2.getString("a"));
		System.out.println("Row 2, Column b: " + rs2.getString("b"));
		System.out.println("Row 2, Column c length: " + rs2.getString("c").length());

	} catch (SQLException ex) {
		System.out.println("SQLException: " + ex.getMessage());
	}

    }
}
```

File: server/conn.go [hunk 1]
```diff
@@ -1546,6 +1553,11 @@ func (cc *clientConn) writeError(ctx context.Context, e error) error {
 // serverStatus, a flag bit represents server information
 // in the packet.
 func (cc *clientConn) writeEOF(serverStatus uint16) error {
+	if cc.capability&mysql.ClientDeprecateEOF > 0 {
+		logutil.Logger(context.Background()).Error(
+			"Attempt to write EOF to client that has the ClientDeprecateEOF flag set", zap.Stack("stack"))
```
Comment (bb7133) [inline server/conn.go:1558]: Why adding this error message?

Comment (dveeden) [reply server/conn.go:1558]: That is not strictly needed, but when `ClientDeprecateEOF` is active this code shouldn't be reached, if it is there is likely a bug, so I added this to make that visible. We could remove this.

Comment (dveeden) [reply server/conn.go:1558]: @bb7133 what do you think? keep this? remove this? Could you suggest other reviewers as well?

Comment (xhebox) [reply server/conn.go:1558]: I think keeping is harmless.

File: parser/mysql/const.go [hunk 2]
```diff
@@ -155,6 +155,9 @@ const (
 	ClientPluginAuth
 	ClientConnectAtts
 	ClientPluginAuthLenencClientData
+	ClientHandleExpiredPasswords
+	ClientSessionTrack
+	ClientDeprecateEOF
```
Comment (xhebox) [inline parser/mysql/const.go:160]: Maybe address morgo's comments in another PR, https://github.com/pingcap/tidb/pull/32119/files#r800834737

File: server/conn.go [hunk 3]
```diff
@@ -2167,7 +2179,7 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
 	return false, cc.flush(ctx)
 }
 
-func (cc *clientConn) writeColumnInfo(columns []*ColumnInfo, serverStatus uint16) error {
+func (cc *clientConn) writeColumnInfo(ctx context.Context, columns []*ColumnInfo, serverStatus uint16) error {
```
Comment (xhebox) [inline server/conn.go:2182]: ctx is not used?

**Review by xhebox — COMMENTED** [review]
Needs release note

**Comment by bb7133** [comment]
I tried this PR and found:

When using MySQL client to login, an weird number of warnings are reported:
```
➜  ~ mysql --version
mysql  Ver 14.14 Distrib 5.7.28, for linux-glibc2.12 (x86_64) using  EditLine wrapper

➜  ~ mysql -u root -h 127.0.0.1 -P 4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 407
Server version: 5.7.25-TiDB-v6.2.0-alpha-30-gd57613c06d TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> use test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed, **25856** warnings
mysql>
```

When using the following Java code and connect to TiDB, an error is reported:
```
import java.sql.Connection;
import java.sql.Timestamp;
import java.sql.DriverManager;
import java.sql.DatabaseMetaData;
import java.sql.PreparedStatement;
import java.sql.Statement;
import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.sql.ResultSet;
import java.sql.SQLException;

import java.math.BigDecimal;

import java.lang.*;

class Prepare {
	public static void test(String url) throws Exception {
	    Connection conn = DriverManager.getConnection(url, "", "");
		DatabaseMetaData meta = conn.getMetaData();
		System.out.println("URL:" + url);
		System.out.println("Driver version: " + meta.getDriverVersion());
		System.out.println("Server version: " + meta.getDatabaseProductVersion());

		Statement s = conn.createStatement();
		int a, b, c, d;
		while (true) {
		try (PreparedStatement ps = conn.prepareStatement("select *,? from TestTable where col2 < ? and col1 in (?, ?)")) {
			a = 111;
			b = 1;
			c = 2;
			d = 22;
			ps.setInt(1, d);
			ps.setInt(2, a);
			ps.setInt(3, b);
			ps.setInt(4, c);
			ResultSet result = ps.executeQuery();
			if (result.next()) {
				System.out.println("Result: " + result.getString(1) + ", " + result.getString(2) + ", " + result.getString(3));
			} else {
				System.out.println("Result: empty.");
			}

			a = 112;
			b = -2;
			c = -5;
			d = 33;
			ps.setInt(1, d);
			ps.setInt(2, a);
			ps.setInt(3, b);
			ps.setInt(4, c);
			result = ps.executeQuery();
			if (result.next()) {
				System.out.println("Result: " + result.getString(1) + ", " + result.getString(2) + ", " + result.getString(3));
			} else {
				System.out.println("Result: empty.");
			}
		}
		}
	}

   public static void main(String[] args) throws Exception {
	    System.setProperty( "oracle.jdbc.Trace", Boolean.TRUE.toString() );
        Class.forName("com.mysql.jdbc.Driver").newInstance();
        test("jdbc:mysql://address=(protocol=tcp)(host=127.0.0.1)(port=4000)(user=root)/test?&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&useSSL=false&rewriteBatchedStatements=true");
    }
}
```

Schema and data:
```
tidb> create table TestTable(col1 int, col2 int);
Query OK, 0 rows affected (0.01 sec)

tidb> insert into TestTable values (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (8, 8), (9, 9);
Query OK, 9 rows affected (0.01 sec)
Records: 9  Duplicates: 0  Warnings: 0
```

```
➜  java sh run.sh
Note: Prepare.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
URL:jdbc:mysql://address=(protocol=tcp)(host=127.0.0.1)(port=4000)(user=root)/test?&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&useSSL=false&rewriteBatchedStatements=true
Driver version: mysql-connector-java-8.0.27 (Revision: e920b979015ae7117d60d72bcc8f077a839cd791)
Server version: 5.7.25-TiDB-v6.2.0-alpha-30-gd57613c06d
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 5 out of bounds for length 5
        at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:381)
        at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:395)
        at com.mysql.cj.protocol.a.ColumnDefinitionReader.unpackField(ColumnDefinitionReader.java:96)
        at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:77)
        at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:40)
        at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1587)
        at com.mysql.cj.ServerPreparedQuery.serverPrepare(ServerPreparedQuery.java:172)
        at com.mysql.cj.jdbc.ServerPreparedStatement.serverPrepare(ServerPreparedStatement.java:644)
        at com.mysql.cj.jdbc.ServerPreparedStatement.<init>(ServerPreparedStatement.java:135)
        at com.mysql.cj.jdbc.ServerPreparedStatement.getInstance(ServerPreparedStatement.java:104)
        at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1630)
        at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1564)
        at Prepare.test(Prepare.java:28)
        at Prepare.main(Prepare.java:65)
```

BTW, For `ArrayIndexOutOfBoundsException` exception, I encountered it in my own PR(https://github.com/pingcap/tidb/pull/32119) but for now I've no idea how to solve it :(

**Comment by dveeden** [comment]
@bb7133 

> ```Database changed, **25856** warnings```

I can reproduce this with MySQL Client v5.7.28 but not with v5.7.29 or v8.0.29. An older version (v5.7.27) also seems to have the same issue. So there must be something in v5.7.29 that changed this.

From [the 5.7.29 release notes](https://dev.mysql.com/doc/relnotes/mysql/5.7/en/news-5-7-29.html):

> Client programs that used the libmysqlclient C client library could exit upon receipt of an OK packet containing malformed session-tracking information. (Bug #29297620, Bug #29630735)

So this does indicate that there was some change. For now it means we need to test with v5.7.28 before merging this.

File: server/conn.go (hunk 3 above)
Comment (dveeden) [reply server/conn.go]: Fixed this in 04c8d575e17b3f3c253eac7f4d2a094ddd88cb0a

**Comment by dveeden** [comment]
@bb7133 

> ```Database changed, **25856** warnings```

Fixed in 427e29a8740d00858bcc7e0a70a841f105555b8d

**Comment by onlyacat** [comment]
> I tried to trigger the issue with Connector/J that was reported, but wasn't able to reproduce the original issue.
> 
> ```java
> package com.pingcap.tidb_demo_java;
> 
> import java.sql.Connection;
> import java.sql.DriverManager;
> import java.sql.SQLException;
> import java.sql.Statement;
> import java.sql.ResultSet;
> 
> public class App 
> {
>     public static void main( String[] args )
>     {
> 	try {
> 		Class.forName("com.mysql.cj.jdbc.Driver").newInstance();
> 	} catch (Exception ex) {
> 		System.out.println("Failed to load driver: " + ex.getMessage());
> 	}
> 
> 	try {
> 		Connection conn = DriverManager.getConnection("jdbc:mysql:loadbalance://127.0.0.1:4000,127.0.0.1:4001?user=root&useServerPrepStmts=true");
> 		Statement stmt = conn.createStatement();
> 		ResultSet rs = stmt.executeQuery("SELECT @@hostname, @@port");
> 		rs.next();
> 		System.out.println("Connected to: " + rs.getString("@@hostname") + ":" + rs.getString("@@port"));
> 
> 		ResultSet rs2 = stmt.executeQuery("SELECT 'foo' a, 'bar' b, REPEAT('x', 100000) c UNION ALL SELECT 'foo', 'bar', REPEAT('y', 10000000)");
> 		rs2.next();
> 		System.out.println("Row 1, Column a: " + rs2.getString("a"));
> 		System.out.println("Row 1, Column b: " + rs2.getString("b"));
> 		System.out.println("Row 1, Column c length: " + rs2.getString("c").length());
> 
> 		rs2.next();
> 		System.out.println("Row 2, Column a: " + rs2.getString("a"));
> 		System.out.println("Row 2, Column b: " + rs2.getString("b"));
> 		System.out.println("Row 2, Column c length: " + rs2.getString("c").length());
> 
> 	} catch (SQLException ex) {
> 		System.out.println("SQLException: " + ex.getMessage());
> 	}
> 
>     }
> }
> ```

Hello dveeden! I am the reporter of the issue with Connector/J :) Usually, this bug is hard to reproduce because some problems are related to the network status.

The response packets should be large enough and the PMTU should be small enough so that the responses will be divided into multiple segments. This issue will happen only when the last segment contains a part of the EOF packet or the total EOF packet.

For the test case, I guess one reason it is not working is that the destination is the loopback(127.0.0.1), in which the PMTU is set to a maximum possible value, as for my machine it is 65535.
```
> ping -s 65535 -M do 127.0.0.1  
Error: packet size 65535 is too large. Maximum is 65507
```

I also submitted a test case for this (mysql/mysql-connector-j#74) before, not sure if it can help you to reproduce the bug or not.

**Comment by dveeden** [comment]
> I tried this PR and found:
> 
> When using MySQL client to login, an weird number of warnings are reported:
> 
> ```
> ➜  ~ mysql --version
> mysql  Ver 14.14 Distrib 5.7.28, for linux-glibc2.12 (x86_64) using  EditLine wrapper
> 
> ➜  ~ mysql -u root -h 127.0.0.1 -P 4000
> Welcome to the MySQL monitor.  Commands end with ; or \g.
> Your MySQL connection id is 407
> Server version: 5.7.25-TiDB-v6.2.0-alpha-30-gd57613c06d TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible
> 
> Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
> 
> Oracle is a registered trademark of Oracle Corporation and/or its
> affiliates. Other names may be trademarks of their respective
> owners.
> 
> Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
> 
> mysql> use test;
> Reading table information for completion of table and column names
> You can turn off this feature to get a quicker startup with -A
> 
> Database changed, **25856** warnings
> mysql>
> ```
> 
> When using the following Java code and connect to TiDB, an error is reported:
> 
> ```
> import java.sql.Connection;
> import java.sql.Timestamp;
> import java.sql.DriverManager;
> import java.sql.DatabaseMetaData;
> import java.sql.PreparedStatement;
> import java.sql.Statement;
> import java.sql.Date;
> import java.sql.Time;
> import java.sql.Timestamp;
> import java.sql.ResultSet;
> import java.sql.SQLException;
> 
> import java.math.BigDecimal;
> 
> import java.lang.*;
> 
> class Prepare {
> 	public static void test(String url) throws Exception {
> 	    Connection conn = DriverManager.getConnection(url, "", "");
> 		DatabaseMetaData meta = conn.getMetaData();
> 		System.out.println("URL:" + url);
> 		System.out.println("Driver version: " + meta.getDriverVersion());
> 		System.out.println("Server version: " + meta.getDatabaseProductVersion());
> 
> 		Statement s = conn.createStatement();
> 		int a, b, c, d;
> 		while (true) {
> 		try (PreparedStatement ps = conn.prepareStatement("select *,? from TestTable where col2 < ? and col1 in (?, ?)")) {
> 			a = 111;
> 			b = 1;
> 			c = 2;
> 			d = 22;
> 			ps.setInt(1, d);
> 			ps.setInt(2, a);
> 			ps.setInt(3, b);
> 			ps.setInt(4, c);
> 			ResultSet result = ps.executeQuery();
> 			if (result.next()) {
> 				System.out.println("Result: " + result.getString(1) + ", " + result.getString(2) + ", " + result.getString(3));
> 			} else {
> 				System.out.println("Result: empty.");
> 			}
> 
> 			a = 112;
> 			b = -2;
> 			c = -5;
> 			d = 33;
> 			ps.setInt(1, d);
> 			ps.setInt(2, a);
> 			ps.setInt(3, b);
> 			ps.setInt(4, c);
> 			result = ps.executeQuery();
> 			if (result.next()) {
> 				System.out.println("Result: " + result.getString(1) + ", " + result.getString(2) + ", " + result.getString(3));
> 			} else {
> 				System.out.println("Result: empty.");
> 			}
> 		}
> 		}
> 	}
> 
>    public static void main(String[] args) throws Exception {
> 	    System.setProperty( "oracle.jdbc.Trace", Boolean.TRUE.toString() );
>         Class.forName("com.mysql.jdbc.Driver").newInstance();
>         test("jdbc:mysql://address=(protocol=tcp)(host=127.0.0.1)(port=4000)(user=root)/test?&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&useSSL=false&rewriteBatchedStatements=true");
>     }
> }
> ```
> 
> Schema and data:
> 
> ```
> tidb> create table TestTable(col1 int, col2 int);
> Query OK, 0 rows affected (0.01 sec)
> 
> tidb> insert into TestTable values (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (8, 8), (9, 9);
> Query OK, 9 rows affected (0.01 sec)
> Records: 9  Duplicates: 0  Warnings: 0
> ```
> 
> ```
> ➜  java sh run.sh
> Note: Prepare.java uses or overrides a deprecated API.
> Note: Recompile with -Xlint:deprecation for details.
> Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
> URL:jdbc:mysql://address=(protocol=tcp)(host=127.0.0.1)(port=4000)(user=root)/test?&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&useSSL=false&rewriteBatchedStatements=true
> Driver version: mysql-connector-java-8.0.27 (Revision: e920b979015ae7117d60d72bcc8f077a839cd791)
> Server version: 5.7.25-TiDB-v6.2.0-alpha-30-gd57613c06d
> Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 5 out of bounds for length 5
>         at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:381)
>         at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:395)
>         at com.mysql.cj.protocol.a.ColumnDefinitionReader.unpackField(ColumnDefinitionReader.java:96)
>         at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:77)
>         at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:40)
>         at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1587)
>         at com.mysql.cj.ServerPreparedQuery.serverPrepare(ServerPreparedQuery.java:172)
>         at com.mysql.cj.jdbc.ServerPreparedStatement.serverPrepare(ServerPreparedStatement.java:644)
>         at com.mysql.cj.jdbc.ServerPreparedStatement.<init>(ServerPreparedStatement.java:135)
>         at com.mysql.cj.jdbc.ServerPreparedStatement.getInstance(ServerPreparedStatement.java:104)
>         at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1630)
>         at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1564)
>         at Prepare.test(Prepare.java:28)
>         at Prepare.main(Prepare.java:65)
> ```
> 
> BTW, For `ArrayIndexOutOfBoundsException` exception, I encountered it in my own PR(#32119) but for now I've no idea how to solve it :(

I also saw the `ArrayIndexOutOfBoundsException` on `v6.2.0-alpha-nightly-20220727` nightly build with Connector/J 8.0.29

File: parser/mysql/const.go (hunk 2 above)
Comment (dveeden) [reply parser/mysql/const.go]: Done in 970508b5ec2612910b65639bda6521d821fd0072

**Comment by dveeden** [comment]
> ```
> ➜  java sh run.sh
> Note: Prepare.java uses or overrides a deprecated API.
> Note: Recompile with -Xlint:deprecation for details.
> Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
> URL:jdbc:mysql://address=(protocol=tcp)(host=127.0.0.1)(port=4000)(user=root)/test?&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&useSSL=false&rewriteBatchedStatements=true
> Driver version: mysql-connector-java-8.0.27 (Revision: e920b979015ae7117d60d72bcc8f077a839cd791)
> Server version: 5.7.25-TiDB-v6.2.0-alpha-30-gd57613c06d
> Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 5 out of bounds for length 5
>         at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:381)
>         at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:395)
>         at com.mysql.cj.protocol.a.ColumnDefinitionReader.unpackField(ColumnDefinitionReader.java:96)
>         at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:77)
>         at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:40)
>         at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1587)
>         at com.mysql.cj.ServerPreparedQuery.serverPrepare(ServerPreparedQuery.java:172)
>         at com.mysql.cj.jdbc.ServerPreparedStatement.serverPrepare(ServerPreparedStatement.java:644)
>         at com.mysql.cj.jdbc.ServerPreparedStatement.<init>(ServerPreparedStatement.java:135)
>         at com.mysql.cj.jdbc.ServerPreparedStatement.getInstance(ServerPreparedStatement.java:104)
>         at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1630)
>         at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1564)
>         at Prepare.test(Prepare.java:28)
>         at Prepare.main(Prepare.java:65)
> ```
> 
> 
>     
>   
> 
> BTW, For `ArrayIndexOutOfBoundsException` exception, I encountered it in my own PR(#32119) but for now I've no idea how to solve it :(

That seems to be coming from this code:
https://github.com/mysql/mysql-connector-j/blob/e920b979015ae7117d60d72bcc8f077a839cd791/src/main/protocol-impl/java/com/mysql/cj/protocol/a/NativePacketPayload.java#L381

**Comment by dveeden** [comment]
I was able to reproduce the `ArrayIndexOutOfBoundsException` with the code from @bb7133. Looking into a fix now.
```
Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
URL:jdbc:mysql://address=(protocol=tcp)(host=127.0.0.1)(port=4000)(user=root)/test?&useUnicode=true&characterEncoding=UTF-8&useServerPrepStmts=true&useSSL=false&rewriteBatchedStatements=true
Driver version: mysql-connector-java-8.0.29-SNAPSHOT (Revision: ${revinfo.commit})
Server version: 5.7.25-TiDB-v6.2.0-alpha-568-g08a58c690
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 5 out of bounds for length 5
	at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:393)
	at com.mysql.cj.protocol.a.NativePacketPayload.readInteger(NativePacketPayload.java:407)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.unpackField(ColumnDefinitionReader.java:96)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:77)
	at com.mysql.cj.protocol.a.ColumnDefinitionReader.read(ColumnDefinitionReader.java:40)
	at com.mysql.cj.protocol.a.NativeProtocol.read(NativeProtocol.java:1651)
	at com.mysql.cj.ServerPreparedQuery.serverPrepare(ServerPreparedQuery.java:163)
	at com.mysql.cj.jdbc.ServerPreparedStatement.serverPrepare(ServerPreparedStatement.java:565)
	at com.mysql.cj.jdbc.ServerPreparedStatement.<init>(ServerPreparedStatement.java:130)
	at com.mysql.cj.jdbc.ServerPreparedStatement.getInstance(ServerPreparedStatement.java:98)
	at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1629)
	at com.mysql.cj.jdbc.ConnectionImpl.prepareStatement(ConnectionImpl.java:1563)
	at com.pingcap.javademo2.Prepare.test(App.java:30)
	at com.pingcap.javademo2.Prepare.main(App.java:67)
```

image

**Comment by dveeden** [comment]
I fixed the `ArrayIndexOutOfBoundsException` issue in 99f740ca26b2cc1acb10a206f883863be8a7bfcd

File: server/conn.go [hunk 4]
```diff
@@ -1507,17 +1507,21 @@ func (cc *clientConn) flush(ctx context.Context) error {
 
 func (cc *clientConn) writeOK(ctx context.Context) error {
 	msg := cc.ctx.LastMessage()
-	return cc.writeOkWith(ctx, msg, cc.ctx.AffectedRows(), cc.ctx.LastInsertID(), cc.ctx.Status(), cc.ctx.WarningCount())
+	return cc.writeOkWith(ctx, msg, cc.ctx.AffectedRows(), cc.ctx.LastInsertID(), cc.ctx.Status(), cc.ctx.WarningCount(), false)
 }
 
-func (cc *clientConn) writeOkWith(ctx context.Context, msg string, affectedRows, lastInsertID uint64, status, warnCnt uint16) error {
+func (cc *clientConn) writeOkWith(ctx context.Context, msg string, affectedRows, lastInsertID uint64, status, warnCnt uint16, eof bool) error {
```
Comment (bb7133) [inline server/conn.go:1513]: IMHO it's a better idea to change the signature to(more straightforward):

```
func (cc *clientConn) writeOkWith(ctx context.Context, header byte, msg string, affectedRows, lastInsertID uint64, status, warnCnt uint16)
```
This is what I did in #32119, what do you think :)

**Review by bb7133 — COMMENTED** [review]
Could you please add some tests in `conn_test.go`?

File: server/conn.go (hunk 4 above)
Comment (dveeden) [reply server/conn.go]: Done in 78772e40fb9ffb5cab0e939a95bc062d9ccbba29

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by bb7133** [comment]
PTAL @xhebox

File: server/conn.go [hunk 5]
```diff
@@ -1507,17 +1507,17 @@ func (cc *clientConn) flush(ctx context.Context) error {
 
 func (cc *clientConn) writeOK(ctx context.Context) error {
 	msg := cc.ctx.LastMessage()
-	return cc.writeOkWith(ctx, msg, cc.ctx.AffectedRows(), cc.ctx.LastInsertID(), cc.ctx.Status(), cc.ctx.WarningCount())
+	return cc.writeOkWith(ctx, msg, cc.ctx.AffectedRows(), cc.ctx.LastInsertID(), cc.ctx.Status(), cc.ctx.WarningCount(), mysql.OKHeader)
 }
 
-func (cc *clientConn) writeOkWith(ctx context.Context, msg string, affectedRows, lastInsertID uint64, status, warnCnt uint16) error {
+func (cc *clientConn) writeOkWith(ctx context.Context, msg string, affectedRows, lastInsertID uint64, status, warnCnt uint16, header byte) error {
```
Comment (xhebox) [inline server/conn.go:1513]: Maybe rename to `writeWithHeader`?

File: server/conn_stmt.go [hunk 6]
```diff
@@ -98,8 +98,10 @@ func (cc *clientConn) handleStmtPrepare(ctx context.Context, sql string) error {
 			}
 		}
 
-		if err := cc.writeEOF(0); err != nil {
-			return err
+		if !(cc.capability&mysql.ClientDeprecateEOF > 0) {
```
Comment (xhebox) [inline server/conn_stmt.go:101]: Why not `cc.capability&mysql.ClientDeprecateEOF  == 0`?

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: server/conn.go (hunk 5 above)
Comment (dveeden) [reply server/conn.go:1513]: You mean `writeOkWithHeader` or `writeWithHeader`?

I think the current name is fine as it is always an `OK_Packet`, but with two kinds of headers (0x00, OK or 0xFE EOF). This is done in a weird way in the protocol, but that's just how it is.

See also:
- https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_ok_packet.html
- https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html

I'd like to keep the current name if that's ok with you

File: server/conn_stmt.go (hunk 6 above)
Comment (dveeden) [reply server/conn_stmt.go:101]: That would be ok as well.
```suggestion
		if !(cc.capability&mysql.ClientDeprecateEOF == 0) {
```

Comment (dveeden) [reply server/conn_stmt.go]: Done in 0abd5e3367a1a0caf6258e1d91eca020e4bdaf89

File: server/conn.go (hunk 5 above)
Comment (xhebox) [reply server/conn.go:1513]: OK

**Review by xhebox — APPROVED** [review]

File: server/conn_stmt.go [hunk 7]
```diff
@@ -98,8 +98,10 @@ func (cc *clientConn) handleStmtPrepare(ctx context.Context, sql string) error {
 			}
 		}
 
-		if err := cc.writeEOF(0); err != nil {
-			return err
+		if cc.capability&mysql.ClientDeprecateEOF == 0 {
```
Comment (shiyuhang0) [inline server/conn_stmt.go:101]: Hi, according to the MySQL doc https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_com_stmt_prepare.html#sect_protocol_com_stmt_prepare_response.

> if num_params > 0 and CLIENT_OPTIONAL_RESULTSET_METADATA is not set or if medatdata_follows is RESULTSET_METADATA_FULL num_params packets will follow. Then a [EOF_Packet](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html) will be transmitted, provided that CLIENT_DEPRECATE_EOF is not set.

It seems we always send num_params packets? why can this work?

Comment (dveeden) [reply server/conn_stmt.go:101]: There is no `CLIENT_OPTIONAL_RESULTSET_METADATA` defined in TiDB yet.

https://github.com/pingcap/tidb/blob/37bd052ee53d1180a89e12b6a01a2ec854f357b6/parser/mysq/const.go#L163

I think the second sentence in that quote is not dependent on the first part.

So maybe read it like:

> if num_params > 0 and CLIENT_OPTIONAL_RESULTSET_METADATA is not set or if medatdata_follows is RESULTSET_METADATA_FULL num_params packets will follow.

> Then a [EOF_Packet](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html) will be transmitted, provided that CLIENT_DEPRECATE_EOF is not set.


And when we exclude `CLIENT_OPTIONAL_RESULTSET_METADATA` and `medatdata_follows` (which requires the former) then it is like this:

> if num_params > 0, num_params packets will follow.

> Then a [EOF_Packet](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html) will be transmitted, provided that CLIENT_DEPRECATE_EOF is not set.

This seems to match what the code does.

Would be good to compare this with MySQL packet traces if you think MySQL is behaving differently.

**Pull request merged**